icu4c/source/test/intltest/msfmrgts.cpp

```diff
@@ -989,21 +989,34 @@ void MessageFormatRegressionTest::Test4112104()
 }
 
 void MessageFormatRegressionTest::TestICU12584() {
-    // 42 placeholders.
     // We were able to handle only 10 due to the bug ICU-12584.
-     UnicodeString pattern("{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}" \
-        "{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}" \
-        "{3}{1}{2}{3}");
+     UnicodeString pattern(
+        u"{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}"
+        u"{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}{1}{2}{3}");
 
     UErrorCode status = U_ZERO_ERROR;
-    int count;
-    const Format **formats;
 
     MessageFormat msg(pattern, status);
-    failure(status, "MessageFormat");
-
-    formats = msg.getFormats(count);
-    logln("Placeholder count: %d", count);
+    failure(status, "Flat message");
+
+    int32_t count;
+    msg.getFormats(count);
+    logln("Plain message placeholder count: %d", static_cast<int>(count));
+
+    // Make sure we iterate only over top level arguments (so we don't over allocate).
+    UnicodeString inner_pattern("{1}{1,select,fem {1} masc {2} other {3}}{2}");
+    status = U_ZERO_ERROR;
+    MessageFormat inner_msg(inner_pattern, status);
+    failure(status, "Inner message");
+
+    count = 0;
+    const int32_t expected_count = 3;
+    inner_msg.getFormats(count);
+    if (count != expected_count) {
+        errln("Error: expected count %d, got %d", static_cast<int>(expected_count), static_cast<int>(count));
```

Comment:
This is fine, but next time please use assertEquals(...)   :-)

Reply:
done

Didn't know about that one, fancy :)